src/bpfprogram.cpp

```diff
@@ -1,569 +1,104 @@
-                         insn_off;
-
-    dstoff += sizeof(struct bpf_func_info);
-    ptr += func_info_rec_size;
-  }
+  bpf_program__set_autoattach(bpf_prog_, false);
 }
 
-namespace {
-/*
- * Searches the verifier's log for err_pattern. If a match is found, extracts
- * the name and ID of the problematic helper and throws a HelperVerifierError.
- *
- * Example verfier log extract:
- *     [...]
- *     36: (b7) r3 = 64                      ; R3_w=64
- *     37: (85) call bpf_d_path#147
- *     helper call is not allowed in probe
- *     [...]
- *
- *  In the above log, "bpf_d_path" is the helper's name and "147" is the ID.
- */
-void maybe_throw_helper_verifier_error(std::string_view log,
-                                       std::string_view err_pattern,
-                                       const std::string &exception_msg_suffix)
+struct bpf_program *BpfProgram::bpf_prog() const
 {
-  auto err_pos = log.find(err_pattern);
-  if (err_pos == log.npos)
-    return;
-
-  std::string_view call_pattern = " call ";
-  auto call_pos = log.rfind(call_pattern, err_pos);
-  if (call_pos == log.npos)
-    return;
-
-  auto helper_begin = call_pos + call_pattern.size();
-  auto hash_pos = log.find("#", helper_begin);
-  if (hash_pos == log.npos)
-    return;
-
-  auto eol = log.find("\n", hash_pos + 1);
-  if (eol == log.npos)
-    return;
-
-  auto helper_name = std::string{ log.substr(helper_begin,
-                                             hash_pos - helper_begin) };
-  auto func_id = std::stoi(
-      std::string{ log.substr(hash_pos + 1, eol - hash_pos - 1) });
-
-  std::string msg = std::string{ "helper " } + helper_name +
-                    exception_msg_suffix;
-  throw HelperVerifierError(msg, static_cast<libbpf::bpf_func_id>(func_id));
+  return bpf_prog_;
 }
-} // namespace
 
-void BpfProgram::load(const Probe &probe,
-                      const BpfBytecode &bytecode,
-                      const BTF &btf,
-                      BPFfeature &feature)
+char *BpfProgram::log_buf() const
 {
-  auto &insns = getCode();
-  auto func_infos = getFuncInfos();
-  const char *license = "GPL";
-  int log_level = 0;
-  std::stringstream logbuf; // while stderr is silenced and for joining lines
-
-  uint64_t log_buf_size = probe.log_size;
-  auto log_buf = std::make_unique<char[]>(log_buf_size);
-  std::string name;
-  std::string tracing_type;
-
-  /*
-   * Fallbacks for kernels that don't cope well with func_infos and BTF, and may
-   * check the program's kernel version. Try witout BTF and func_infos and when
-   * needed, also pass the kernel version (try three different methods):
-   */
-  struct load_strategy {
-    bool btf;
-    bool func_infos;
-    KernelVersionMethod kver;
-  } load_configs[] = {
-    { .btf = true, .func_infos = true, .kver = None },  // BTF and func_infos
-    { .btf = true, .func_infos = false, .kver = None }, // BTF, no func_infos
-    /* Fallbacks for Linux <5.0: No BTF and func_infos, with kernel version: */
```

Comment:
@viktormalik do you happen know to whether the old kernel(e.g 4.19) still work w/o these fallbacks code?

Reply:
@simpleton now all program loading is handled by libbpf so if there's a sufficiently new libbpf on the system, it *should* work.

Reply:
Anything older than 5.4 we don't support (https://github.com/bpftrace/bpftrace/blob/master/docs/dependency_support.md#linux-kernel).